tests/lib/test_grpc.cpp

```diff
@@ -0,0 +1,909 @@
+ *
+ * Copyright (c) 2021, LabN Consulting, L.L.C
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <time.h>
+#include <zebra.h>
+
+#include "filter.h"
+#include "frr_pthread.h"
+#include "libfrr.h"
+#include "routing_nb.h"
+#include "northbound_cli.h"
+#include "thread.h"
+#include "vrf.h"
+#include "vty.h"
+
+#include "staticd/static_debug.h"
+#include "staticd/static_nb.h"
+#include "staticd/static_vrf.h"
+#include "staticd/static_vty.h"
+#include "staticd/static_zebra.h"
+
+// GRPC C++ includes
+#include <string>
+#include <sstream>
+#include <grpc/grpc.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+#include "grpc/frr-northbound.grpc.pb.h"
+
+DEFINE_HOOK(frr_late_init, (struct thread_master * tm), (tm));
+DEFINE_KOOH(frr_fini, (), ());
+
+struct vty *vty;
+
+bool mpls_enabled;
+struct thread_master *master;
+struct zebra_privs_t static_privs = {0};
+struct frrmod_runtime *grpc_module;
+
+extern const char *json_expect1;
+extern const char *json_expect2;
+extern const char *json_expect3;
+extern const char *json_loadconf1;
+
+int test_dbg = 1;
+
+void inline test_debug(const std::string &s)
+{
+	if (test_dbg)
+		std::cout << s << std::endl;
+}
+
+// static struct option_chain modules[] = {{ .arg = "grpc:50051" }]
+// static struct option_chain **modnext = modules->next;
+
+static const struct frr_yang_module_info *const staticd_yang_modules[] = {
+	&frr_interface_info, &frr_filter_info, &frr_routing_info,
+	&frr_staticd_info,   &frr_vrf_info,
+};
+
+static int grpc_thread_stop(struct thread *thread);
+
+static void static_startup(void)
+{
+	// struct frrmod_runtime module;
+	// static struct option_chain *oc;
+	char moderr[256] = {};
+	cmd_init(1);
+
+	zlog_aux_init("NONE: ", LOG_DEBUG);
+	zprivs_preinit(&static_privs);
+	zprivs_init(&static_privs);
+
+	/* Load the server side module */
+	grpc_module = frrmod_load("grpc:50051", "../lib/.libs", moderr,
+				  sizeof(moderr));
+	if (!grpc_module)
+		grpc_module = frrmod_load("grpc:50051", "../../lib/.libs",
+					  moderr, sizeof(moderr));
+	if (!grpc_module)
+		grpc_module = frrmod_load("grpc:50051", "../../../lib/.libs",
+					  moderr, sizeof(moderr));
```

Comment:
lmao what is going on here

Reply:
It's called, "WTF is the built in-tree grpc module?" I'm very much open to some other way of locating the thing.

Reply:
I think I can actually count on argv[0] being valid path to the binary (one hopes the unit-tests are being found using "PATH"), so I can probably get the path that way.. maybe even a function to locate the base "tests" dir would be useful.

Reply:
ok cleaned up to use exe path to find locally built module